app/src/org/commcare/connect/network/parser/CompleteProfileResponseParser.java

```diff
@@ -0,0 +1,25 @@
+package org.commcare.connect.network.parser;
+
+import org.commcare.android.database.connect.models.PersonalIdSessionData;
+import org.commcare.utils.JsonExtensions;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/**
+ * Parses a JSON response from the complete profile API call
+ * and populates a PersonalIdSessionData instance.
+ */
+public class CompleteProfileResponseParser implements PersonalIdApiResponseParser{
+    /**
+     * Parses and sets values on the given PersonalIdSessionData instance.
+     *
+     * @param sessionData the instance to populate
+     * @throws JSONException if a parsing error occurs
+     */
+    @Override
+    public void parse(JSONObject json, PersonalIdSessionData sessionData) throws JSONException {
+        sessionData.setPersonalId(JsonExtensions.optStringSafe(json, "username", null));
+        sessionData.setDbKey(JsonExtensions.optStringSafe(json, "db_key", null));
+        sessionData.setOauthPassword(JsonExtensions.optStringSafe(json, "password", null));
```

Comment:
these keys must be mandatory now right ?

Reply:
think using `optStringSafe` is fine there but we should do a separate invalidation later on each of these paramertes to be non null and throw a different exception than Json Exception, maybe `IllegalStateException` . This is because we still want to check for "null" strings when validating this, the handling for which is only built in `optStringSafe` and not `getString`